Interactivity Router: Update for latest GB changes.

sirreal opened the pull request:
Accounts for changes to the `@wordpress/interactivity-router` module in:

- https://github.com/WordPress/gutenberg/pull/65123
- https://github.com/WordPress/gutenberg/pull/65539

Static localized strings are passed via script module data passing instead of in Interactivity API store state.
Redundant HTML used for `aria-live` regions is removed. This functionality is not handled by the `@wordpress/a11y` script module added in #7405.

Trac ticket: https://core.trac.wordpress.org/ticket/60647

## Testing

Testing this requires package updates from https://github.com/WordPress/gutenberg/pull/65539. It anticipates the package being released and updated in Core.

(Copied from https://github.com/WordPress/gutenberg/pull/65539)

> This PR is difficult to test on its own because it requires the package to be updated as a Core dependency for proper testing. It does not affect Gutenberg behavior, only the package's behavior in Core. There are other ways of doing it but this works:
> 
> - Get a checkout of Core from https://github.com/WordPress/wordpress-develop/pull/7304. This is the build that will be running.
> - Run `npm ci` in the Core checkout to install packages.
> - Build the package from this PR (`npm run build:packages` in Gutenberg).
> - Replace the `@wordpress/interactivity-router` package in Core's `node_modules` (`node_modules/@wordpress/interactivity-router`) with the package directory in Gutenberg (remove the directory and copy the directory from Gutenberg `packages/interactivity-router`).
> - In Core, run the npm build script you use, in my case `npm run dev`.
> 
> Then test out that the interactivity router a11y functionality is working. A good way to test is the query block with "force page reload" option disabled. Specifically, the aria-live regions of the page should be updated (with localized text if in non-English locale) announcing "Page loaded."

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

sirreal committed:
Add interactivity-router translations via script module data

swissspidy commented on src/wp-includes/interactivity-api/class-wp-interactivity-api.php:
```diff
@@ -199,6 +199,26 @@ public function print_client_interactivity_data() {
+	 * @param array $data Data to filter.
+	 * @return array Data for the Interactivity Router script module.
+	 */
+	public function filter_script_module_interactivity_router_data( array $data ): array {
```
Can't we find a better way?

This is really a step backwards from the JS I18N mechanism we implemented all those years ago 😞 

Translators loose important context because the strings are no longer co-located with the code where they're being used. This also fails as soon as soon as you are dealing with plurals.

sirreal replied:
This comment and conversation continued in https://github.com/WordPress/gutenberg/pull/65123#discussion_r1754292031.

sirreal committed:
Remove iAPI-specific screen reader text

This reverts commit c6fcf75e5ea98c8148ff1b46aa8bf61e98ac451d.

sirreal committed:
Remove localized strings from state

sirreal commented:
This is building on https://github.com/WordPress/wordpress-develop/pull/7405.

sirreal commented:
See https://github.com/WordPress/gutenberg/pull/65539

sirreal committed:
Remove interactivity router screen reader test

sirreal committed:
Revert "Remove localized strings from state"

This reverts commit 2d8401fbe376bc77fb5c51d8991b61e6b27ca2c6.

sirreal committed:
Restore some server-side init

sirreal commented on src/wp-includes/interactivity-api/class-wp-interactivity-api.php:
```diff
@@ -1016,16 +1031,16 @@ private function data_wp_router_region_processor( WP_Interactivity_API_Directive
-							'loaded'  => __( 'Page Loaded.' ),
-						),
-					),
+					'navigation' => new stdClass(),
```
Maybe this is better?
```suggestion
					'navigation' => array(
						'hasStarted'  => false,
						'hasFinished' => false,
					),
```

There's a bug where if `navigation` is not defined on the server, interactivity-router may error.

michalczaplinski replied:
To confirm my understanding: `	'navigation' => new stdClass()` currently fixes that bug, right?

sirreal replied:
Yes, this ensures the `navigation` is an object (`{}`) when serialized to JavaScript. Then there's no failure when it assigns to `navigation.hasStarted` .

michalczaplinski reviewed (approved):
Looks good to me @sirreal ! 👍 

I've tested it following the instructions in the PR.

michalczaplinski commented on src/wp-includes/interactivity-api/class-wp-interactivity-api.php:
```diff
@@ -990,12 +1011,6 @@ class="wp-interactivity-router-loading-bar"
 				data-wp-class--start-animation="state.navigation.hasStarted"
 				data-wp-class--finish-animation="state.navigation.hasFinished"
 			></div>
-			<div
```
Sorry to bring this up after having already approved!

I've just noticed that we should rename the function to `print_router_loading_markup()`, update its docstring and update the reference where we pass this function to the filter [over here](https://github.com/WordPress/wordpress-develop/pull/7304/files#diff-ba219807876e59a40e6ccd29916a088f32a9951d2210209c15cba58c0a337657R1053).

sirreal replied:
This is a public method, extenders could be calling it or modifying its behavior in filters. Renaming the method could be disruptive. It's true the method name isn't a perfect description now. What do you think of adding a note in the method documentation instead of changing the name?

michalczaplinski replied:
Maybe we could do this:
- Create a new `print_router_markup` function
- Deprecate the “old” method, and have it log a deprecation.
- The “old” method calls  the new `print_router_markup` function.

I think that might be the way to go.

I’m looking at some existing functions/methods that have the @deprecated docstring. For example this function is deprecated, calls `_deprecated_function()` and then calls the updated version:

https://github.com/WordPress/wordpress-develop/blob/9e290835d6f006997d59ebaf063df113bb15a363/src/wp-includes/class-wp-scripts.php#L479-L505

sirreal replied:
Ok, I've searched for the method name and it only appears in Gutenberg across searchable [plugins](https://wpdirectory.net/search/01J8QN3A1M1XEH7HVE1E3G3109) and [themes](https://wpdirectory.net/search/01J8QN3B0VC884R6D4AACRQ5QC).

Let's just rename it.

michalczaplinski replied:
All right!

sirreal committed:
Rename method to reflect its new behavior

The print_router_loading_and_screen_reader_markup method name was very
specific about its behavior which has changed. Rename the method to a
more descriptive name and update its documentation.

michalczaplinski reviewed (approved):
Looks good to me! 👍

michalczaplinski commented:
Committed with https://core.trac.wordpress.org/changeset/59097

sirreal commented:
This was reverted to avoid removing the public method.

New PR with these changes: https://github.com/WordPress/wordpress-develop/pull/7447
